### PR TITLE
feat(content-sharing): Display contacts in collaborator form

### PR DESCRIPTION
**Pull request by sahiga**
Display contacts in the "Invite people" form in the ContentSharing Element.

Filtered contacts:
content-sharing-display-users-autofill

List of roles for the new collaborator:
content-sharing-display-users-roles

### CODE REVIEW

File: src/api/Users.js [hunk 1]
```diff
@@ -36,6 +36,19 @@ class Users extends Base {
...
+     * API URL for fetching all users in the current user's enterprise
+     *
+     * @param {string} filterTerm Filter for enterprise users
+     * @return {string} URL for fetching enterprise users
+     */
+    getUsersInEnterpriseUrl(filterTerm: string): string {
+        if (!filterTerm) {
+            throw new Error('Missing filter term');
+        }
+        return `${this.getBaseApiUrl()}/users?filter_term=${filterTerm}`;
```
Comment (booc0mtaco) [inline src/api/Users.js:49]: Perhaps we need go get in the habit of using [`query-string`](https://www.npmjs.com/package/query-string) to handle these cases. this will automatically handled escaping of `filterTerm` where needed (and if desired).

We already have this package as a dependency.

File: src/features/unified-share-modal/utils/convertData.js [hunk 2]
```diff
@@ -233,3 +233,27 @@ export const convertCollabsResponse = (
...
+ * @param {UserCollection} contactsAPIData
+ * @param {string|null} currentUserID
+ */
+export const convertContactsResponse = (
+    contactsAPIData: UserCollection,
+    currentUserID: string | null,
+): Array<contactType> => {
+    const { entries } = contactsAPIData;
+
+    if (!entries || !entries.length) return [];
```
Comment (booc0mtaco) [inline src/features/unified-share-modal/utils/convertData.js:249]: you should be able to apply `.filter` to the return statement below so you avoid having multiple returns in this. More functional that way, and fewer branches to test.

File: src/api/Users.js [hunk 3]
```diff
@@ -36,6 +36,19 @@ class Users extends Base {
...
     }
 
+    /**
+     * API URL for fetching all users in the current user's enterprise
+     *
+     * @param {string} filterTerm Filter for enterprise users
+     * @return {string} URL for fetching enterprise users
+     */
+    getUsersInEnterpriseUrl(filterTerm: string): string {
+        if (!filterTerm) {
```
Comment (booc0mtaco) [inline src/api/Users.js:46]: Hm, after looking at this, looks like we treat `''` as a possible filter term by default on LN123? It that's the case, when using the default value, it will throw?

Comment (sahiga) [reply src/api/Users.js]: @booc0mtaco Thanks for catching this. I updated the PR based on your comments -- could you take another look?

File: src/api/Users.js [hunk 4]
```diff
@@ -36,6 +36,23 @@ class Users extends Base {
         return `${this.getBaseApiUrl()}/users/${id}/avatar`;
     }
 
+    /**
+     * API URL for fetching all users in the current user's enterprise
+     *
+     * @param {string} [filterTerm] Optional filter for enterprise users
+     * @return {string} URL for fetching enterprise users
```
Comment (booc0mtaco) [inline src/api/Users.js:43]: `@returns`

File: src/features/unified-share-modal/utils/convertData.js [hunk 5]
```diff
@@ -233,3 +233,25 @@ export const convertCollabsResponse = (
 
     return { collaborators };
 };
+
+/**
+ * Convert an enterprise users API response into an array of internal USM contacts.
+ *
+ * @param {UserCollection} contactsAPIData
+ * @param {string|null} currentUserID
+ */
```
Comment (booc0mtaco) [inline src/features/unified-share-modal/utils/convertData.js:242]: `@returns {[contactType]}` or whatever the JSDoc syntax for this is

File: src/api/Base.js [hunk 6]
```diff
@@ -170,11 +170,13 @@ class Base {
      * Generic success handler
      *
      * @param {Object} data - The response data
+     * @returns {Promise|null|undefined}
      */
-    successHandler = (data: any): void => {
+    successHandler = (data: any): Promise<any> | null | typeof undefined => {
```
Comment (ConradJChan) [inline src/api/Base.js:175]: This change to have the success handler return a promise seems not ideal to me. It seems like we should adhere to a callback pattern or a Promise based pattern rather than a hybrid of the two. Is this change necessary?

File: src/elements/content-sharing/SharingNotification.js [hunk 7]
```diff
@@ -239,6 +248,43 @@ function SharingNotification({
         setCollaboratorsList,
     ]);
 
+    // Set the getContacts function, which is used for inviting collaborators in the USM
+    React.useEffect(() => {
```
Comment (ConradJChan) [inline src/elements/content-sharing/SharingNotification.js:252]: tbh, having these large `useEffect` calls make it a bit hard to read the `SharingNotification` component. Would it make sense to create custom hooks that could be reused by other components and simplify this component? Maybe a `useContacts` and `useCollaborators` hooks?

File: src/features/unified-share-modal/utils/convertData.js [hunk 8]
```diff
@@ -233,3 +233,25 @@ export const convertCollabsResponse = (
...
+export const convertContactsResponse = (
+    contactsAPIData: UserCollection,
+    currentUserID: string | null,
+): Array<contactType> => {
+    const { entries = [] } = contactsAPIData;
+
+    // Return all users except for the current user
+    return entries.reduce((filteredContacts, contact) => {
```
Comment (ConradJChan) [inline src/features/unified-share-modal/utils/convertData.js:250]: This could also be written in a more functional style like `entries.filter(...).map(...)`

File: src/api/Base.js (hunk 6 above)
Comment (sahiga) [reply src/api/Base.js:175]: Yeah, I agree that it's not ideal. `getContactsPromise()` in `ContactsField` (https://github.com/box/box-ui-elements/blob/master/src/features/unified-share-modal/ContactsField.js#L114) uses Promises, but the existing flow uses callbacks. To get the response to show up properly in `ContactsField`, I returned the result of `successHandler()`. Do you know if there are other ways to handle this case?

File: src/elements/content-sharing/SharingNotification.js (hunk 7 above)
Comment (sahiga) [reply src/elements/content-sharing/SharingNotification.js:252]: Good idea. Will look into refactoring this component.

File: src/api/Base.js (hunk 6 above)
Comment (ConradJChan) [reply src/api/Base.js:175]: Could we do something like what the [Feed.js](https://github.com/box/box-ui-elements/blob/4858ac4bd28c0f7a49b547f300f8145a9b5c01f3/src/api/Feed.js#L1261) does? If `updatedGetContactsFn` needs to return a promise, it could create a `new Promise(...)` and pass that `resolve` into the success and error callbacks.

**Comment by sahiga** [comment]
@booc0mtaco @ConradJChan Thanks for reviewing :) I updated the PR; could you take another look?

File: src/elements/content-sharing/ContentSharing.js [hunk 9]
```diff
@@ -92,7 +99,7 @@ function ContentSharing({ apiHost, displayInModal, itemID, itemType, language, t
         setChangeSharedLinkAccessLevel(null);
         setChangeSharedLinkPermissionLevel(null);
         setCollaboratorsList(null);
-        setCurrentUserID(null);
+        // setCurrentUserID(null);
```
Comment (swfree) [inline src/elements/content-sharing/ContentSharing.js:102]: just checking this is an intentional change?

File: src/elements/content-sharing/hooks/useCollaborators.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+ * @param {API} api
+ * @param {string} itemID
+ * @param {ItemType} itemType
+ * @param {string} ownerEmail
+ * @param {boolean} isCurrentUserOwner
+ * @param {Function} [handleSuccess]
+ * @param {Function} [handleError]
+ * @returns {collaboratorsListType | null}
+ */
+function useCollaborators(
```
Comment (swfree) [inline src/elements/content-sharing/hooks/useCollaborators.js:27]: now that these hooks are separated out, do we need unit tests for them or are they covered by the tests in ContentSharing.test.js?

Comment (sahiga) [reply src/elements/content-sharing/hooks/useCollaborators.js:27]: They are covered by ContentSharing.test.js and SharingNotification.test.js, but I'll add more tests specifically for the hooks.

File: src/elements/content-sharing/hooks/useCollaborators.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+): collaboratorsListType | null {
+    const [collaboratorsList, setCollaboratorsList] = React.useState<collaboratorsListType | null>(null);
+
+    React.useEffect(() => {
+        if (collaboratorsList) return;
+
+        const handleGetCollaborationsSuccess = (response: Collaborations) => {
+            const updatedCollaboratorsList = convertCollabsResponse(response, ownerEmail, isCurrentUserOwner);
+            setCollaboratorsList(updatedCollaboratorsList);
+            if (handleSuccess) {
```
Comment (ConradJChan) [inline src/elements/content-sharing/hooks/useCollaborators.js:44]: could `handle*` be defaulted to `noop` and avoid these if checks?

File: src/elements/content-sharing/hooks/useContacts.js [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+        const updatedGetContactsFn: GetContactsFnType = () => (filterTerm: string) => {
+            return new Promise((resolve: (result: Array<contactType>) => void) => {
+                api.getUsersAPI(false).getUsersInEnterprise(
+                    itemID,
+                    (response: UserCollection) => {
+                        if (handleSuccess) {
+                            handleSuccess();
+                        }
+                        return resolve(convertContactsResponse(response, currentUserID));
```
Comment (ConradJChan) [inline src/elements/content-sharing/hooks/useContacts.js:41]: Should this massaging of data be done outside of the hook to keep this more general purpose?

File: src/elements/content-sharing/hooks/useContacts.js [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    React.useEffect(() => {
+        if (getContacts) return;
+
+        const updatedGetContactsFn: GetContactsFnType = () => (filterTerm: string) => {
+            return new Promise((resolve: (result: Array<contactType>) => void) => {
+                api.getUsersAPI(false).getUsersInEnterprise(
+                    itemID,
+                    (response: UserCollection) => {
+                        if (handleSuccess) {
+                            handleSuccess();
```
Comment (ConradJChan) [inline src/elements/content-sharing/hooks/useContacts.js:39]: Should the `handleSuccess` return the data?

File: src/elements/content-sharing/SharingNotification.js [hunk 14]
```diff
@@ -179,65 +194,23 @@ function SharingNotification({
...
+        null,
+        () => handleError(contentSharingMessages.collaboratorsLoadingError),
+    );
+    setCollaboratorsList(collaboratorsList);
+
+    // Set the getContacts function
+    const getContactsFn: GetContactsFnType | null = useContacts(api, currentUserID, itemID, null, () =>
+        handleError(contentSharingMessages.getContactsError),
+    );
+    setGetContacts(() => getContactsFn);
```
[the remark this answers is not in the document]
Comment (sahiga) [reply src/elements/content-sharing/SharingNotification.js:213]: Yup, I think so. Without the wrapper, the state gets reset automatically and causes an infinite rendering loop.

**Comment by sahiga** [comment]
@swfree @ConradJChan - could you review again? 🙏

File: src/elements/content-sharing/SharingNotification.js [hunk 15]
```diff
@@ -199,16 +202,20 @@ function SharingNotification({
...
         itemType,
-        ownerEmail,
-        currentUserID === ownerID,
-        null,
+        data => convertCollabsResponse(data, ownerEmail, currentUserID === ownerID),
         () => handleError(contentSharingMessages.collaboratorsLoadingError),
     );
-    setCollaboratorsList(collaboratorsList);
+    if (collaboratorsList && !collabsExist) {
+        setCollaboratorsList(collaboratorsList);
```
Comment (ConradJChan) [inline src/elements/content-sharing/SharingNotification.js:209]: Typically, I haven't seen our `successCallback`s return data (L205) -- could we call `setCollaboratorsList(convertCollabsResponse(...))` here?

File: src/elements/content-sharing/hooks/useCollaborators.js [hunk 16]
```diff
@@ -28,29 +26,21 @@ function useCollaborators(
...
     React.useEffect(() => {
         if (collaboratorsList) return;
 
         const handleGetCollaborationsSuccess = (response: Collaborations) => {
-            const updatedCollaboratorsList = convertCollabsResponse(response, ownerEmail, isCurrentUserOwner);
-            setCollaboratorsList(updatedCollaboratorsList);
-            if (handleSuccess) {
-                handleSuccess();
-            }
+            setCollaboratorsList(handleSuccess(response));
```
Comment (ConradJChan) [inline src/elements/content-sharing/hooks/useCollaborators.js:38]: Should we rely on the `handleSuccess` to return the data? What about something like:
```
setCollaboratorsList(response);
handleSuccess(response);
```

File: src/elements/content-sharing/SharingNotification.js [hunk 17]
```diff
@@ -198,14 +199,15 @@ function SharingNotification({
     ]);
 
     // Set the collaborators list
-    const collaboratorsList: collaboratorsListType | null = useCollaborators(
-        api,
-        itemID,
-        itemType,
-        data => convertCollabsResponse(data, ownerEmail, currentUserID === ownerID),
-        () => handleError(contentSharingMessages.collaboratorsLoadingError),
+    const collaboratorsListFromAPI: Collaborations | null = useCollaborators(api, itemID, itemType, undefined, () =>
```
Comment (ConradJChan) [inline src/elements/content-sharing/SharingNotification.js:202]: Thoughts on having an options argument for the custom hook instead? That way you can avoid having to pass the `undefined` for the `successCallback`

File: src/elements/content-sharing/SharingNotification.js [hunk 18]
```diff
@@ -198,14 +199,15 @@ function SharingNotification({
...
+        handleError(contentSharingMessages.collaboratorsLoadingError),
     );
-    if (collaboratorsList && !collabsExist) {
+    if (collaboratorsListFromAPI && !collabsExist) {
```
Comment (ConradJChan) [inline src/elements/content-sharing/SharingNotification.js:205]: Is the `collabsExist` state needed? Is the truthiness of `collaboratorsListFromAPI` insufficient

File: src/elements/content-sharing/SharingNotification.js (hunk 17 above)
Comment (sahiga) [reply src/elements/content-sharing/SharingNotification.js]: Good call. Fixed

File: src/elements/content-sharing/SharingNotification.js (hunk 18 above)
Comment (sahiga) [reply src/elements/content-sharing/SharingNotification.js]: Yeah, this prevents an infinite loop in which `setCollaboratorsList()` keeps getting called. I updated the component to check for the existence of state values from the parent component.
